storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,535 @@
+func New(provided Config) (storage.PeerStore, error) {
+	cfg := provided.Validate()
+
+	etcdClient, err := clientv3.New(cfg.EtcdConfig)
+	if err != nil {
+		return nil, err
+	}
+	etcdClient.KV = namespace.NewKV(etcdClient.KV, cfg.EtcdNamespace)
+	etcdClient.Watcher = namespace.NewWatcher(etcdClient.Watcher, cfg.EtcdNamespace)
+	etcdClient.Lease = namespace.NewLease(etcdClient.Lease, cfg.EtcdNamespace)
+
+	memoryStore, err := memory.New(cfg.MemoryStore)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := &peerStore{
+		cfg:         cfg,
+		etcd:        etcdClient,
+		memoryStore: memoryStore,
+		closed:      make(chan struct{}),
+	}
+
+	if err = ps.Bootstrap(); err != nil {
+		return nil, err
+	}
+
+	return ps, nil
+}
+
+type serializedPeer []byte
+
+func newPeerKey(p *bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+	return b
+}
+
+const (
+	EtcdKeySep 					= "/"
+	EtcdAnnouncePrefix 			= "announce"
+	EtcdAnnounceSeederPrefix 	= "s" 				//len 1
+	EtcdAnnounceLeecherPrefix	= "l"				//len 1
+)
+
+func newAnnounceKey(ih *bittorrent.InfoHash, p *bittorrent.Peer, seeder bool) ([]byte) {
+	prefix, sep := []byte(EtcdAnnouncePrefix), []byte(EtcdKeySep)
+	var peerType []byte
+	if seeder {
+		peerType = []byte(EtcdAnnounceSeederPrefix)
+	} else {
+		peerType = []byte(EtcdAnnounceLeecherPrefix)
+	}
+
+	var pk serializedPeer
+	if p != nil {
+		pk = newPeerKey(p)
+	}
+
+	key := make([]byte, 0, len(prefix) + len(sep) * 3 + len(ih) + len(peerType) + len(pk))
+	key = append(key, prefix...)
+	key = append(key, sep...)
+	key = append(key, ih[:]...)
+	key = append(key, sep...)
+	key = append(key, peerType...)
+	key = append(key, sep...)
+	key = append(key, pk...)
+
+	return key
+}
+
+func decodeAnnounceKey(key []byte) (*bittorrent.InfoHash, []byte, bool) {
+	prefix, sep := []byte(EtcdAnnouncePrefix), []byte(EtcdKeySep)
+
+	offset := len(prefix) + len(sep)
+	ih := bittorrent.InfoHashFromBytes(key[offset:offset+20])
+
+	offset += len(ih) + len(sep)
+	peerType := key[offset:offset + 1]
+	var seeder bool
+	if string(peerType) == EtcdAnnounceSeederPrefix {
+		seeder = true
+	} else {
+		seeder = false
+	}
+
+	offset += 1 + len(sep)
+	peerKey := key[offset:]
+
+	return &ih, peerKey, seeder
+}
+
+type peerStore struct {
+	cfg         Config
+	etcd        *clientv3.Client
+	memoryStore storage.PeerStore
+	closed      chan struct{}
+}
+
+var _ storage.PeerStore = &peerStore{}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.PutSeeder(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdPutOrUpdatePeer(ih, p, true); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdPutOrUpdatePeer(ih, p, true); err != nil {
```

Comment:
As mentioned in PR comment (first one), not worried about etcd failures since announces are served from in-memory. inconsistency results in missing announces on tracker reboot but no functionality loss.
Now that I think about it, if someone choses to do sync operation for etcd, they are ok with perf hit, so yes might as well update etcd before in-memory op.
Will fix.

Reply:
In fact, this whole logic is repeated in every route handler. ugly right now. I have become comfortable with lambdas in golang so that opens new avenues of cool refactoring